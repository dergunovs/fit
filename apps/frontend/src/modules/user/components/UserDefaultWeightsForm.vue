<template>
  <div>
    <UiFlex v-if="exercisesToChooseWeight.length > 0" column data-test="user-default-weights">
      <UiFlex v-for="exercise in exercisesToChooseWeight" :key="exercise._id" column data-test="user-default-weight">
        <div data-test="user-default-weight-title">{{ exercise[localeField('title', locale)] }}</div>

        <UiSelect
          v-if="exercise._id"
          :modelValue="props.modelValue?.[exercise._id]"
          :options="exercise.options"
          @update:modelValue="(value) => updateWeights(Number(value), exercise._id)"
          :lang="locale"
          data-test="user-default-weight-select"
        />
      </UiFlex>
    </UiFlex>

    <div v-else data-test="user-default-weights-empty">{{ t('exercise.noAvailable') }}</div>
  </div>
</template>

<script setup lang="ts">
import { computed } from 'vue';
import { useI18n } from 'vue-i18n';
import { IExercise, IUserDefaultWeights, IUserEquipment } from 'fitness-tracker-contracts';
import { UiFlex, UiSelect } from 'mhz-ui';
import { localeField } from 'mhz-helpers';

import { getExercisesToChooseDefaultWeight } from '@/exercise/helpers';

interface IProps {
  userEquipments: IUserEquipment[];
  exercises: IExercise[];
  modelValue?: IUserDefaultWeights;
}

interface IEmit {
  'update:modelValue': [value?: IUserDefaultWeights];
}

const props = defineProps<IProps>();
const emit = defineEmits<IEmit>();

const { t, locale } = useI18n();

const exercisesToChooseWeight = computed(() =>
  getExercisesToChooseDefaultWeight(props.exercises, props.userEquipments)
);

function updateWeights(weight: number, id?: string) {
  if (!id) return;

  const weightsToUpdate = props.modelValue ? { ...props.modelValue, [id]: weight } : { [id]: weight };

  emit('update:modelValue', weightsToUpdate);
}
</script>
