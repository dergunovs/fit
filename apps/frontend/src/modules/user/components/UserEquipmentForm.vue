<template>
  <UiFlex gap="16" column>
    <UiFlex>
      <UiFlex column>
        <div>Оборудование</div>

        <UiFlex>
          <UiSelect
            v-model="choosenEquipment"
            :options="equipmentsFiltered"
            :isDisabled="isEditEquipment"
            lang="ru"
            data-test="user-form-equipment-options"
          />
        </UiFlex>
      </UiFlex>

      <UiFlex v-if="choosenEquipment?.isWeights" column>
        <div>Возможный вес</div>

        <UiFlex>
          <UiInput
            v-model="choosenEquipmentWeightInput"
            type="number"
            step="1"
            min="1"
            max="500"
            data-test="user-form-weights"
          />

          <UiButton
            :isDisabled="isAddWeightDisabled"
            layout="secondary"
            isNarrow
            @click="addWeight"
            data-test="user-form-add-weight"
          >
            Добавить
          </UiButton>
        </UiFlex>
      </UiFlex>
    </UiFlex>

    <UiFlex v-if="choosenEquipmentWeights.length" column>
      <div>Добавленные веса</div>

      <UiFlex>
        <UiChip v-for="weight in choosenEquipmentWeights" :key="weight">
          <IconWeight width="16" height="16" /><span>{{ weight }}</span> кг.

          <button
            type="button"
            @click="deleteWeight(weight)"
            :class="$style.delete"
            data-test="user-form-delete-weight"
          >
            ×
          </button>
        </UiChip>
      </UiFlex>
    </UiFlex>

    <UiFlex>
      <UiButton v-if="isEditEquipment" @click="saveEquipment" data-test="user-form-save-equipment">
        Сохранить оборудование
      </UiButton>

      <UiButton v-if="isEditEquipment" @click="resetEquipment" layout="secondary" data-test="user-form-reset-equipment">
        Отменить редактирование
      </UiButton>

      <UiButton v-else :isDisabled="isAddEquipmentDisabled" @click="addEquipment" data-test="user-form-add-equipment">
        Добавить оборудование
      </UiButton>
    </UiFlex>

    <UiFlex v-if="props.modelValue?.length" column>
      <div>Добавленное оборудование</div>

      <UiFlex wrap>
        <UiChip
          v-for="equipment in props.modelValue"
          :key="`${equipment.equipment?._id}-${equipment.weights?.join()}`"
          data-test="user-form-equipment"
        >
          <span data-test="user-form-equipment-title">{{ equipment.equipment?.title }}</span>

          <span v-for="weight in equipment.weights" :key="weight" data-test="user-form-equipment-weights">
            <span data-test="user-form-equipment-weight">{{ weight }}</span> кг.
          </span>

          <button
            v-if="equipment.equipment?.isWeights"
            type="button"
            @click="editEquipment(equipment)"
            :class="$style.edit"
            data-test="user-form-equipment-edit"
          >
            <IconEdit width="20" height="20" />
          </button>

          <button
            v-if="equipment.equipment?.title"
            type="button"
            @click="deleteEquipment(equipment.equipment.title)"
            :class="$style.delete"
            data-test="user-form-equipment-delete"
          >
            ×
          </button>
        </UiChip>
      </UiFlex>
    </UiFlex>
  </UiFlex>
</template>

<script setup lang="ts">
import { ref, computed } from 'vue';
import { IEquipment, IUserEquipment } from 'fitness-tracker-contracts';
import { UiButton, UiFlex, UiInput, UiSelect, UiChip } from 'mhz-ui';

import IconWeight from '@/common/icons/weight.svg';
import IconEdit from '@/common/icons/edit.svg';

interface IProps {
  equipments: IEquipment[];
  modelValue?: IUserEquipment[];
}

const props = defineProps<IProps>();
const emit = defineEmits<{ 'update:modelValue': [value?: IUserEquipment[]] }>();

const choosenEquipment = ref<IEquipment>();
const choosenEquipmentWeightInput = ref<number>(0);
const choosenEquipmentWeights = ref<number[]>([]);

const isEditEquipment = ref(false);

const equipmentsFiltered = computed(() =>
  props.equipments.filter(
    (equipment) =>
      !props.modelValue?.some((equipmentToFilter) => equipment.title === equipmentToFilter.equipment?.title)
  )
);

const isAddWeightDisabled = computed(
  () =>
    choosenEquipmentWeightInput.value === 0 ||
    choosenEquipmentWeights.value.includes(Number(choosenEquipmentWeightInput.value))
);

const isAddEquipmentDisabled = computed(
  () =>
    !choosenEquipment.value ||
    (choosenEquipment.value.isWeights && !choosenEquipmentWeights.value.length) ||
    props.modelValue?.some((equipment) => equipment.equipment?.title === choosenEquipment.value?.title)
);

function addWeight() {
  if (isAddWeightDisabled.value) return;

  choosenEquipmentWeights.value = [...choosenEquipmentWeights.value, Number(choosenEquipmentWeightInput.value)].sort();
}

function deleteWeight(weightToDelete: number) {
  choosenEquipmentWeights.value = choosenEquipmentWeights.value.filter((weight) => weight !== weightToDelete);
}

function addEquipment() {
  if (isAddEquipmentDisabled.value) return;

  const updatedEquipments = props.modelValue?.length
    ? [...props.modelValue, { equipment: choosenEquipment.value, weights: choosenEquipmentWeights.value }]
    : [{ equipment: choosenEquipment.value, weights: choosenEquipmentWeights.value }];

  emit('update:modelValue', updatedEquipments);

  resetEquipment();
}

function editEquipment(equipment: IUserEquipment) {
  isEditEquipment.value = true;

  choosenEquipment.value = equipment.equipment;
  choosenEquipmentWeightInput.value = 0;
  choosenEquipmentWeights.value = equipment.weights?.length ? [...equipment.weights] : [];
}

function saveEquipment() {
  const updatedEquipments = props.modelValue?.map((equipment) => {
    if (equipment.equipment?.title === choosenEquipment.value?.title) {
      return { equipment: choosenEquipment.value, weights: choosenEquipmentWeights.value };
    } else return equipment;
  });

  emit('update:modelValue', updatedEquipments);

  resetEquipment();
}

function deleteEquipment(title: string) {
  const updatedEquipments = props.modelValue?.filter((equipment) => equipment.equipment?.title !== title);

  emit('update:modelValue', updatedEquipments);
}

function resetEquipment() {
  choosenEquipment.value = undefined;
  choosenEquipmentWeightInput.value = 0;
  choosenEquipmentWeights.value = [];

  isEditEquipment.value = false;
}
</script>

<style module lang="scss">
.edit {
  display: flex;
  align-items: center;
  justify-content: center;
  padding: 2px;
  cursor: pointer;
  background: none;
  border: none;
}

.delete {
  display: flex;
  align-items: center;
  justify-content: center;
  width: 20px;
  height: 20px;
  font-size: 0.75rem;
  font-weight: 700;
  color: var(--color-white);
  cursor: pointer;
  background-color: var(--color-error);
  border: none;
  border-radius: 50%;
}
</style>
