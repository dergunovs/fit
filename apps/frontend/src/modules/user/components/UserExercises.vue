<template>
  <UiFlex column gap="16">
    <UiFlex wrap>
      <UiChip v-for="exercise in props.exercises" :key="exercise._id" align="center" data-test="user-exercises">
        <span data-test="user-exercise-title">{{ exercise[localeField('title', locale)] }}</span>

        <button
          v-if="exercise._id"
          @click="emit('edit', exercise)"
          type="button"
          :class="$style.edit"
          data-test="user-exercise-edit"
        >
          <IconEdit width="20" height="20" />
        </button>
      </UiChip>
    </UiFlex>
  </UiFlex>
</template>

<script setup lang="ts">
import { useI18n } from 'vue-i18n';
import { UiFlex, UiChip } from 'mhz-ui';
import { localeField } from 'mhz-helpers';
import { IExercise } from 'fitness-tracker-contracts';

import IconEdit from '@/common/icons/edit.svg';

interface IProps {
  exercises?: IExercise[];
}

interface IEmit {
  edit: [exercise: IExercise];
}

const props = defineProps<IProps>();
const emit = defineEmits<IEmit>();

const { locale } = useI18n();
</script>

<style module lang="scss">
.edit {
  display: flex;
  align-items: center;
  justify-content: center;
  padding: 2px;
  cursor: pointer;
  background: none;
  border: none;
}
</style>
