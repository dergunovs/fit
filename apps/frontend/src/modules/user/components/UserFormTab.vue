<template>
  <UiFlex column gap="16">
    <h3 data-test="user-form-tab-title">{{ props.title }}</h3>

    <p data-test="user-form-tab-description">{{ props.description }}</p>

    <slot></slot>
  </UiFlex>
</template>

<script setup lang="ts">
import { UiFlex } from 'mhz-ui';

interface IProps {
  title: string;
  description: string;
}

const props = defineProps<IProps>();
</script>
