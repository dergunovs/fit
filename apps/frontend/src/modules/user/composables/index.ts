import { computed, Ref } from 'vue';

import { useTI18n } from '@/common/composables';

export function useUserFormTabs(isEdit: boolean, isAdmin: Ref<boolean>) {
  const { t } = useTI18n();

  const generalTab = computed(() => {
    return { value: 'general', title: t('user.general') };
  });

  const exercisesTab = computed(() => {
    return { value: 'exercises', title: t('exercise.many') };
  });

  const otherTabs = computed(() => [
    { value: 'goals', title: t('goals') },
    { value: 'weight', title: t('weight') },
    { value: 'templates', title: t('template.many') },
  ]);

  const TABS = computed(() => {
    if (!isEdit) return [generalTab.value];
    if (isAdmin.value) return [generalTab.value, ...otherTabs.value];

    return [generalTab.value, ...otherTabs.value, exercisesTab.value];
  });

  return { TABS };
}
