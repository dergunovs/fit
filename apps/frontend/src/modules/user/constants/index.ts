export const URL_USER = '/users';
export const URL_USER_CREATE = `${URL_USER}/create`;
export const URL_USER_EDIT = `${URL_USER}/edit`;
export const URL_USER_PROFILE = `${URL_USER}/profile`;

export const USER_GOALS_OPTIONS = {
  activities: [1, 2, 3, 4, 5],
  sets: [12, 16, 20, 24, 30],
  repeats: [8, 10, 12, 15, 20],
  duration: [20, 30, 40, 50, 60],
};
