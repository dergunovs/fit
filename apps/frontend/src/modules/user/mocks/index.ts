import { vi } from 'vitest';
import {
  TDeleteUserDTO,
  TGetUsersDTO,
  TPostUserDataDTO,
  TPostUserDTO,
  TUpdateUserDataDTO,
  TUpdateUserDTO,
} from 'fitness-tracker-contracts';

import { userService } from '@/user/services';
import { mockMutationReply, mockQueryReply } from '@/common/mocks';
import { USER_FIXTURE, USERS_FIXTURE } from '@/user/fixtures';
import { IOnSuccess } from '@/common/interface';

const spyGetUser = vi.spyOn(userService, 'getOne').mockReturnValue(mockQueryReply(USER_FIXTURE));

const mockGetUsersData: TGetUsersDTO = { data: USERS_FIXTURE, total: USERS_FIXTURE.length };

const spyGetUsers = vi.spyOn(userService, 'getMany').mockImplementation(() => mockQueryReply(mockGetUsersData));

const spyCreateUser = vi.fn();
const spyUpdateUser = vi.fn();
const spyDeleteUser = vi.fn();

const mockOnSuccess: IOnSuccess = {
  create: undefined,
  update: undefined,
  delete: undefined,
};

vi.spyOn(userService, 'create').mockImplementation((options: { onSuccess?: () => Promise<void> }) => {
  if (options.onSuccess) mockOnSuccess.create = options.onSuccess;

  return mockMutationReply<TPostUserDTO, TPostUserDataDTO>(spyCreateUser);
});

vi.spyOn(userService, 'update').mockImplementation((options: { onSuccess?: () => Promise<void> }) => {
  if (options.onSuccess) mockOnSuccess.update = options.onSuccess;

  return mockMutationReply<TUpdateUserDTO, TUpdateUserDataDTO>(spyUpdateUser);
});

vi.spyOn(userService, 'delete').mockImplementation((options: { onSuccess?: () => Promise<void> }) => {
  if (options.onSuccess) mockOnSuccess.delete = options.onSuccess;

  return mockMutationReply<TDeleteUserDTO, string>(spyDeleteUser);
});

export { spyGetUser, spyGetUsers, spyCreateUser, spyUpdateUser, spyDeleteUser, mockOnSuccess, mockGetUsersData };
