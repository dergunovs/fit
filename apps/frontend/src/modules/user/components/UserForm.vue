<template>
  <div>
    <UiFlex @submit.prevent="submit" tag="form" column gap="24" align="flex-start" data-test="user-form">
      <div v-if="isAdmin" :class="$style.admin" data-test="user-form-admin">Пользователь с правами администратора</div>

      <UiField label="Электронная почта" isRequired :error="error('email')">
        <UiInput v-model="formData.email" data-test="user-form-email" />
      </UiField>

      <UiField label="Имя" isRequired :error="error('name')">
        <UiInput v-model="formData.name" data-test="user-form-name" />
      </UiField>

      <UiField v-if="!props.user?._id" label="Пароль" isRequired :error="error('password')">
        <UiInput v-model="formData.password" data-test="user-form-password" />
      </UiField>

      <UserEquipmentForm
        v-if="equipments"
        :equipments="equipments"
        v-model="formData.equipments"
        data-test="user-form-equipments"
      />

      <UserDefaultWeightsForm
        v-if="formData.equipments && exercises"
        :userEquipments="formData.equipments"
        :exercises="exercises"
        v-model="formData.defaultWeights"
        data-test="user-form-default-weights"
      />

      <FormButtons
        :id="props.user?._id"
        :isLoading="isLoadingPost || isLoadingUpdate"
        @delete="(id) => mutateDelete(id)"
        data-test="user-form-buttons"
      />
    </UiFlex>
  </div>
</template>

<script setup lang="ts">
import { ref, computed, onMounted } from 'vue';
import { useRouter } from 'vue-router';
import { UiField, UiFlex, UiInput, toast } from 'mhz-ui';
import { useQueryClient, useValidator, required, email, clone } from 'mhz-helpers';
import { API_ACTIVITY_STATISTICS, API_AUTH_GET, API_USER, IUser } from 'fitness-tracker-contracts';

import FormButtons from '@/common/components/FormButtons.vue';
import UserEquipmentForm from '@/user/components/UserEquipmentForm.vue';
import UserDefaultWeightsForm from '@/user/components/UserDefaultWeightsForm.vue';

import { URL_USER } from '@/user/constants';
import { userService } from '@/user/services';
import { equipmentService } from '@/equipment/services';
import { exerciseService } from '@/exercise/services';

interface IProps {
  user?: IUser | null;
}

const props = defineProps<IProps>();

const router = useRouter();

const queryClient = useQueryClient();

const { data: equipments } = equipmentService.getAll();

const { data: exercises } = exerciseService.getAll();

const formData = ref<IUser>({
  email: '',
  name: '',
  password: '',
  equipments: [],
  defaultWeights: {},
  role: 'user',
});

const isAdmin = computed(() => props.user?.role === 'admin');

const { mutate: mutatePost, isPending: isLoadingPost } = userService.create({
  onSuccess: async () => {
    await queryClient.refetchQueries({ queryKey: [API_USER] });
    toast.success('Пользователь добавлен');
    router.push(URL_USER);
  },
});

const { mutate: mutateUpdate, isPending: isLoadingUpdate } = userService.update({
  onSuccess: async () => {
    await queryClient.refetchQueries({ queryKey: [API_USER] });
    await queryClient.refetchQueries({ queryKey: [API_AUTH_GET] });
    await queryClient.refetchQueries({ queryKey: [API_ACTIVITY_STATISTICS] });
    toast.success('Пользователь обновлен');
  },
});

const { mutate: mutateDelete } = userService.delete({
  onSuccess: async () => {
    queryClient.removeQueries({ queryKey: [API_USER] });
    await queryClient.refetchQueries({ queryKey: [API_USER] });
    toast.success('Пользователь удален');
    router.push(URL_USER);
  },
});

const rules = computed(() => {
  return {
    email: [required('ru'), email('ru')],
    name: required('ru'),
    password: !props.user?._id && required('ru'),
  };
});

const { error, isValid } = useValidator(formData, rules);

function submit() {
  if (!isValid()) return;

  if (props.user?._id) {
    mutateUpdate(formData.value);
  } else {
    mutatePost(formData.value);
  }
}

onMounted(() => {
  if (props.user) formData.value = clone(props.user);
});
</script>

<style module lang="scss">
.admin {
  font-weight: 700;
  color: var(--color-success);
}
</style>
