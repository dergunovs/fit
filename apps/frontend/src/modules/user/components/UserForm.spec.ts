import { nextTick } from 'vue';
import { describe, it, expect, beforeEach, afterEach } from 'vitest';
import { VueWrapper, enableAutoUnmount } from '@vue/test-utils';
import { API_ACTIVITY_STATISTICS, API_AUTH_GET, API_USER, IUserEquipment } from 'fitness-tracker-contracts';
import { dataTest, deleteAuthHeader } from 'mhz-helpers';
import { UiTabs } from 'mhz-ui';

import UserForm from './UserForm.vue';
import UserEquipmentForm from './UserEquipmentForm.vue';
import UserDefaultWeightsForm from './UserDefaultWeightsForm.vue';
import UserExercises from './UserExercises.vue';
import UserFormTab from './UserFormTab.vue';
import ExerciseForm from '@/exercise/components/ExerciseForm.vue';
import FormButtons from '@/common/components/FormButtons.vue';

import { wrapperFactory } from '@/common/test';
import { USER_FIXTURE } from '@/user/fixtures';
import { mockOnSuccess, spyCreateUser, spyUpdateUser, spyDeleteUser, spyUpdateUserPassword } from '@/user/mocks';
import {
  spyRefetchQueries,
  spyRemoveQueries,
  spyRouterPush,
  spyToastSuccess,
  mockIsValid,
  spyLogout,
} from '@/common/mocks';
import { URL_USER } from '@/user/constants';
import { spyGetEquipments } from '@/equipment/mocks';
import { EQUIPMENTS_FIXTURE } from '@/equipment/fixtures';
import { spyGetExercisesAll, spyGetExercisesCustom } from '@/exercise/mocks';
import { EXERCISE_FIXTURE_CUSTOM, EXERCISES_FIXTURE } from '@/exercise/fixtures';
import { URL_HOME } from '@/common/constants';
import { TOKEN_NAME } from '@/auth/constants';
import { setAdmin } from '@/auth/composables';

const EMAIL = 'unique@mail.ru';
const NAME = 'Уникум';
const PASSWORD = 'unique';
const EQUIPMENTS: IUserEquipment[] = [];

const form = dataTest('user-form');
const formTabs = dataTest('user-form-tabs');
const formTab = dataTest('user-form-tab');
const formEmail = dataTest('user-form-email');
const formName = dataTest('user-form-name');
const formPassword = dataTest('user-form-password');
const formNewPassword = dataTest('user-form-new-password');
const formNewPasswordSpoiler = dataTest('user-form-new-password-spoiler');
const formSetNewPassword = dataTest('user-form-set-new-password');
const formEquipments = dataTest('user-form-equipments');
const formDefaultWeights = dataTest('user-form-default-weights');
const formButtons = dataTest('user-form-buttons');
const formExercises = dataTest('user-form-exercises');
const formAddExercise = dataTest('user-form-add-exercise');
const formExerciseForm = dataTest('user-form-exercise-form');
const formExerciseFormModal = dataTest('user-form-exercise-form-modal');

const wrapperWithUser: VueWrapper<InstanceType<typeof UserForm>> = wrapperFactory(
  UserForm,
  { user: USER_FIXTURE },
  { UserFormTab }
);

let wrapper: VueWrapper<InstanceType<typeof UserForm>>;

beforeEach(() => {
  wrapper = wrapperFactory(UserForm, {}, { UserFormTab });
});

enableAutoUnmount(afterEach);

describe('UserForm', async () => {
  it('exists', async () => {
    expect(wrapper.findComponent(UserForm)).toBeTruthy();
  });

  it('matches snapshot', async () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('shows password field only when creating users', async () => {
    expect(wrapper.find(formPassword).exists()).toBe(true);

    await wrapper.setProps({ user: USER_FIXTURE });

    expect(wrapper.find(formPassword).exists()).toBe(false);

    await wrapper.setProps({ user: undefined });
  });

  it('shows set new password spoiler only when user exists', async () => {
    expect(wrapper.find(formNewPasswordSpoiler).exists()).toBe(false);

    await wrapper.setProps({ user: USER_FIXTURE });

    expect(wrapper.find(formNewPasswordSpoiler).exists()).toBe(true);

    await wrapper.setProps({ user: undefined });
  });

  it('uses validation', async () => {
    mockIsValid.value = false;

    await wrapper.find(form).trigger('submit');

    expect(spyCreateUser).toBeCalledTimes(0);

    mockIsValid.value = true;
  });

  it('creates user', async () => {
    expect(spyCreateUser).toBeCalledTimes(0);
    expect(spyRefetchQueries).toBeCalledTimes(0);
    expect(spyToastSuccess).toBeCalledTimes(0);
    expect(spyRouterPush).toBeCalledTimes(0);

    await wrapper.findComponent(formEmail).setValue(EMAIL);
    await wrapper.findComponent(formName).setValue(NAME);
    await wrapper.findComponent(formPassword).setValue(PASSWORD);

    await wrapper.find(form).trigger('submit');

    expect(spyCreateUser).toBeCalledTimes(1);
    expect(spyCreateUser).toBeCalledWith({
      email: EMAIL,
      name: NAME,
      password: PASSWORD,
      equipments: EQUIPMENTS,
      defaultWeights: {},
      role: 'user',
    });

    await mockOnSuccess.create?.();

    expect(spyRefetchQueries).toBeCalledTimes(1);
    expect(spyRefetchQueries).toBeCalledWith({ queryKey: [API_USER] });

    expect(spyToastSuccess).toBeCalledTimes(1);

    expect(spyRouterPush).toBeCalledTimes(1);
    expect(spyRouterPush).toBeCalledWith(URL_USER);
  });

  it('updates user', async () => {
    expect(spyUpdateUser).toBeCalledTimes(0);
    expect(spyRefetchQueries).toBeCalledTimes(0);
    expect(spyToastSuccess).toBeCalledTimes(0);

    const NEW_NAME = 'Олег';

    await wrapperWithUser.findComponent(formName).setValue(NEW_NAME);

    await wrapperWithUser.find(form).trigger('submit');

    expect(spyUpdateUser).toBeCalledTimes(1);
    expect(spyUpdateUser).toBeCalledWith({
      _id: USER_FIXTURE._id,
      email: USER_FIXTURE.email,
      name: NEW_NAME,
      password: USER_FIXTURE.password,
      equipments: USER_FIXTURE.equipments,
      defaultWeights: USER_FIXTURE.defaultWeights,
      role: USER_FIXTURE.role,
    });

    await mockOnSuccess.update?.();

    expect(spyRefetchQueries).toBeCalledTimes(3);
    expect(spyRefetchQueries).toBeCalledWith({ queryKey: [API_USER] });
    expect(spyRefetchQueries).toBeCalledWith({ queryKey: [API_AUTH_GET] });
    expect(spyRefetchQueries).toBeCalledWith({ queryKey: [API_ACTIVITY_STATISTICS] });

    expect(spyToastSuccess).toBeCalledTimes(1);
  });

  it('updates password', async () => {
    expect(spyUpdateUserPassword).toBeCalledTimes(0);
    expect(spyToastSuccess).toBeCalledTimes(0);

    const NEW_PASSWORD = '123456';

    await wrapperWithUser.findComponent(formNewPassword).setValue(NEW_PASSWORD);

    await wrapperWithUser.find(formSetNewPassword).trigger('click');

    expect(spyUpdateUserPassword).toBeCalledTimes(1);
    expect(spyUpdateUserPassword).toBeCalledWith({ password: NEW_PASSWORD, id: USER_FIXTURE._id });

    await mockOnSuccess.updatePassword?.();

    expect(spyToastSuccess).toBeCalledTimes(1);
  });

  it('deletes user', async () => {
    expect(spyDeleteUser).toBeCalledTimes(0);
    expect(spyRemoveQueries).toBeCalledTimes(0);
    expect(spyRefetchQueries).toBeCalledTimes(0);
    expect(spyToastSuccess).toBeCalledTimes(0);
    expect(spyLogout).toBeCalledTimes(0);

    wrapperWithUser.findComponent<typeof FormButtons>(formButtons).vm.$emit('delete', USER_FIXTURE._id);

    expect(spyDeleteUser).toBeCalledTimes(1);
    expect(spyDeleteUser).toBeCalledWith(USER_FIXTURE._id);

    await mockOnSuccess.delete?.();

    expect(spyRemoveQueries).toBeCalledTimes(1);
    expect(spyRemoveQueries).toBeCalledWith({ queryKey: [API_USER] });

    expect(spyRefetchQueries).toBeCalledTimes(1);
    expect(spyRefetchQueries).toBeCalledWith({ queryKey: [API_USER] });

    expect(spyToastSuccess).toBeCalledTimes(1);

    expect(spyLogout).toBeCalledTimes(1);
    expect(spyLogout).toBeCalledWith(URL_HOME, deleteAuthHeader, TOKEN_NAME);
  });

  it('deletes user with no logout when admin', async () => {
    setAdmin(true);

    expect(spyLogout).toBeCalledTimes(0);
    expect(spyRouterPush).toBeCalledTimes(0);

    wrapperWithUser.findComponent<typeof FormButtons>(formButtons).vm.$emit('delete', USER_FIXTURE._id);

    expect(spyDeleteUser).toBeCalledTimes(1);

    await mockOnSuccess.delete?.();

    expect(spyLogout).toBeCalledTimes(0);

    expect(spyRouterPush).toBeCalledTimes(1);
    expect(spyRouterPush).toBeCalledWith(URL_USER);
  });

  it('sets form buttons id', async () => {
    expect(wrapper.find(formButtons).attributes('id')).toBe(undefined);
    expect(wrapperWithUser.find(formButtons).attributes('id')).toBe(USER_FIXTURE._id);
  });

  it('gets and sets equipment options', async () => {
    expect(spyGetEquipments).toBeCalledTimes(1);
    expect(wrapper.findComponent<typeof UserEquipmentForm>(formEquipments).props('equipments')).toStrictEqual(
      EQUIPMENTS_FIXTURE
    );
  });

  it('gets and sets default weights and exercises', async () => {
    expect(spyGetExercisesAll).toBeCalledTimes(1);

    expect(
      wrapperWithUser.findComponent<typeof UserDefaultWeightsForm>(formDefaultWeights).props('exercises')
    ).toStrictEqual(EXERCISES_FIXTURE);

    expect(
      wrapperWithUser.findComponent<typeof UserDefaultWeightsForm>(formDefaultWeights).props('userEquipments')
    ).toStrictEqual(USER_FIXTURE.equipments);
  });

  it('gets and sets user custom exercises', async () => {
    expect(spyGetExercisesCustom).toBeCalledTimes(1);
    expect(wrapper.findComponent<typeof UserExercises>(formExercises).props('exercises')).toStrictEqual([
      EXERCISE_FIXTURE_CUSTOM,
    ]);
  });

  it('shows form modal by add exercise button click', async () => {
    expect(wrapper.find(formExerciseFormModal).attributes('modelvalue')).toBe('false');

    await wrapper.find(formAddExercise).trigger('click');

    expect(wrapper.find(formExerciseFormModal).attributes('modelvalue')).toBe('true');
  });

  it('sets emited exercise to edit in modal', async () => {
    expect(wrapper.find(formExerciseFormModal).attributes('modelvalue')).toBe('false');

    expect(wrapper.findComponent<typeof ExerciseForm>(formExerciseForm).props('exercise')).toStrictEqual(undefined);

    wrapper.findComponent<typeof UserExercises>(formExercises).vm.$emit('edit', EXERCISE_FIXTURE_CUSTOM);

    await nextTick();

    expect(wrapper.find(formExerciseFormModal).attributes('modelvalue')).toBe('true');

    expect(wrapper.findComponent<typeof ExerciseForm>(formExerciseForm).props('exercise')).toStrictEqual(
      EXERCISE_FIXTURE_CUSTOM
    );
  });

  it('hides form modal and clears current exercise by hide emit', async () => {
    expect(wrapper.find(formExerciseFormModal).attributes('modelvalue')).toBe('false');

    wrapper.findComponent<typeof UserExercises>(formExercises).vm.$emit('edit', EXERCISE_FIXTURE_CUSTOM);

    await nextTick();

    expect(wrapper.find(formExerciseFormModal).attributes('modelvalue')).toBe('true');

    expect(wrapper.findComponent<typeof ExerciseForm>(formExerciseForm).props('exercise')).toStrictEqual(
      EXERCISE_FIXTURE_CUSTOM
    );

    wrapper.findComponent<typeof ExerciseForm>(formExerciseForm).vm.$emit('hide');

    await nextTick();

    expect(wrapper.find(formExerciseFormModal).attributes('modelvalue')).toBe('false');

    expect(wrapper.findComponent<typeof ExerciseForm>(formExerciseForm).props('exercise')).toStrictEqual(undefined);
  });

  it('shows only one tab in user create form', async () => {
    expect(wrapper.findAll(formTab)[0].isVisible()).toBe(true);
    expect(wrapper.findAll(formTab)[1].isVisible()).toBe(false);

    expect(wrapperWithUser.findAll(formTab)[0].isVisible()).toBe(true);
    expect(wrapperWithUser.findAll(formTab)[1].isVisible()).toBe(true);
  });

  it('shows tabs for admin user', async () => {
    setAdmin(false);

    await nextTick();

    expect(wrapperWithUser.findComponent<typeof UiTabs>(formTabs).props('tabs').length).toBe(4);

    setAdmin(true);

    await nextTick();

    expect(wrapperWithUser.findComponent<typeof UiTabs>(formTabs).props('tabs').length).toBe(3);
  });
});
