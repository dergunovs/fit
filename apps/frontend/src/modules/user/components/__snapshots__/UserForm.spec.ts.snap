// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`UserForm > matches snapshot 1`] = `
"<div>
  <div tag="form" column="" gap="16" data-test="user-form">
    <!--v-if-->
    <ui-tabs-stub tabs="[object Object]" modelvalue="general" data-test="user-form-tabs"></ui-tabs-stub>
    <div column="" gap="16" data-test="user-form-tab">
      <h3 data-test="user-form-tab-title">Общее</h3>
      <fieldset label="Имя" isrequired=""><input type="text" modelvalue="" data-test="user-form-name"></fieldset>
      <fieldset label="Электронная почта" isrequired=""><input type="email" modelvalue="" data-test="user-form-email"></fieldset>
      <fieldset label="Пароль" isrequired=""><input type="text" modelvalue="" ispassword="" data-test="user-form-password"></fieldset>
      <!--v-if-->
    </div>
    <div column="" gap="16" data-test="user-form-tab" style="display: none;">
      <h3 data-test="user-form-tab-title">Оборудование</h3>
      <user-equipment-form-stub equipments="[object Object],[object Object]" modelvalue="" data-test="user-form-equipments"></user-equipment-form-stub>
    </div>
    <div column="" gap="16" data-test="user-form-tab" style="display: none;">
      <h3 data-test="user-form-tab-title">Выбор веса</h3>
      <user-default-weights-form-stub userequipments="" exercises="[object Object],[object Object],[object Object]" modelvalue="[object Object]" data-test="user-form-default-weights"></user-default-weights-form-stub>
    </div>
    <div column="" gap="16" data-test="user-form-tab" style="display: none;">
      <h3 data-test="user-form-tab-title">Упражнения</h3>
      <user-exercises-stub exercises="[object Object]" data-test="user-form-exercises"></user-exercises-stub><button data-test="user-form-add-exercise">Добавить своё упражнение</button>
      <div modelvalue="false" width="360" data-test="user-form-exercise-form-modal">
        <div class="_modal_0ded47">
          <h3>Добавить своё упражнение</h3>
          <exercise-form-stub isedit="false" isdisableredirect="true" isfixed="false" data-test="user-form-exercise-form"></exercise-form-stub>
        </div>
      </div>
    </div>
    <form-buttons-stub isedit="false" isemitcancel="false" isfixed="true" data-test="user-form-buttons"></form-buttons-stub>
  </div>
</div>"
`;
